File: POA/Page Replacement/FIFO.c
#include <stdio.h>

#define n 4
int frames[n];
int pages[14] = {7, 0, 1, 2, 0, 3, 0, 4, 2, 3, 0, 3, 2, 3};

void main()
{
    int pagefound,hit=0,miss=0,k=0;
    for(int i=0;i<n;i++)
    {
        frames[i]=-1;
    }

    for(int i=0;i<14;i++)
    {
        pagefound=0;
        for(int j=0;j<n;j++)
        {
            if(frames[j]==pages[i])
            {
                hit++;
                pagefound=1;
                break;
            }
        }
        if(!pagefound)
        {
            miss++;
            frames[k]=pages[i];
            k=(k+1)%n;
        }

        for(int j=0;j<n;j++)
        {
            printf("%d",frames[j]);
        }
        printf("\n");
    }

     printf("Hits:%d , Miss:%d", hit, miss);
}

File: POA/Page Replacement/py/FIFO.py
frameno = int(input("Enter total number of frames: "))
pageno = int(input("Enter total number of pages: "))
pages = []
frames = [-1] * frameno
hit = 0
miss = 0

for i in range(pageno):
    pages.append(int(input("Enter page: ")))

print("Pages:", pages)

k = 0
for i in range(len(pages)):
    if pages[i] in frames:
        hit += 1
        print(f"Page {pages[i]}: Hit")
    else:
        miss += 1
        frames[k] = pages[i]
        print(f"Page {pages[i]} loaded into frame {k}")
        k = (k + 1) % frameno

print("Final frames:", frames)
print("Total hits:", hit)
print("Total misses:", miss)
